colosseum/declaration.py

```diff
@@ -9,22 +10,83 @@
     FLEX_SHRINK_CHOICES, FLEX_START, FLEX_WRAP_CHOICES, FLOAT_CHOICES,
     GRID_AUTO_CHOICES, GRID_AUTO_FLOW_CHOICES, GRID_GAP_CHOICES,
     GRID_PLACEMENT_CHOICES, GRID_TEMPLATE_AREA_CHOICES, GRID_TEMPLATE_CHOICES,
-    INLINE, JUSTIFY_CONTENT_CHOICES, LETTER_SPACING_CHOICES, LTR,
-    MARGIN_CHOICES, MAX_SIZE_CHOICES, MIN_SIZE_CHOICES, NORMAL, NOWRAP,
-    ORDER_CHOICES, ORPHANS_CHOICES, OVERFLOW_CHOICES, PADDING_CHOICES,
-    PAGE_BREAK_AFTER_CHOICES, PAGE_BREAK_BEFORE_CHOICES,
+    INITIAL, INLINE, INVERT, JUSTIFY_CONTENT_CHOICES, LETTER_SPACING_CHOICES,
+    LTR, MARGIN_CHOICES, MAX_SIZE_CHOICES, MEDIUM, MIN_SIZE_CHOICES, NORMAL,
+    NOWRAP, ORDER_CHOICES, ORPHANS_CHOICES, OUTLINE_COLOR_CHOICES,
+    OUTLINE_STYLE_CHOICES, OUTLINE_WIDTH_CHOICES, OVERFLOW_CHOICES,
+    PADDING_CHOICES, PAGE_BREAK_AFTER_CHOICES, PAGE_BREAK_BEFORE_CHOICES,
     PAGE_BREAK_INSIDE_CHOICES, POSITION_CHOICES, QUOTES_CHOICES, ROW,
     SEPARATE, SHOW, SIZE_CHOICES, STATIC, STRETCH, TABLE_LAYOUT_CHOICES,
     TEXT_ALIGN_CHOICES, TEXT_DECORATION_CHOICES, TEXT_INDENT_CHOICES,
     TEXT_TRANSFORM_CHOICES, TOP, TRANSPARENT, UNICODE_BIDI_CHOICES,
     VISIBILITY_CHOICES, VISIBLE, WHITE_SPACE_CHOICES, WIDOWS_CHOICES,
     WORD_SPACING_CHOICES, Z_INDEX_CHOICES, OtherProperty,
-    TextAlignInitialValue, default, INITIAL,
+    TextAlignInitialValue, default,
 )
+from .exceptions import ValidationError
+from .wrappers import Outline
 
 _CSS_PROPERTIES = set()
 
 
+def validated_shorthand_property(name, parser, storage_class):
+    """Define the shorthand CSS font property."""
+
+    def getter(self):
+        # This is the only place we use the storage class as a convenience for the user
+        # and to correctly display the string representation
+        shorthand = storage_class()
+        for property_name in storage_class.properties():
+            try:
+                shorthand[property_name] = getattr(self, '_%s' % property_name)
+            except AttributeError:
+                pass
+
+        if shorthand:
+            return shorthand
+        else:
+            # It no property has been defined for the outline, return initial values
+            for property_name in storage_class.properties():
+                try:
+                    shorthand[property_name] = getattr(self, '%s' % property_name)
+                except AttributeError:
+                    pass
+
+            return shorthand
```

Comment:
This is definitely heading in the right direction, but I'd suggest a slight restructuring.

This implementation hinges on two Python features:
1. `len()` as an evaluation of "truthiness" of the shorthand object
2. `__get/__setitem__` as a way to set attributes.

However, they both strike me as slightly odd abstractions. What's the "length" of a font declaration? Why does a font declaration allow arbitrary index retrival, rather than specific listed properties (i.e., `font['size']`, not `font.size`)?

My suggestion: use a dictionary, rather than the shorthand object as an accumulator. This means the "was anything manually set" check is an "if {}" check.  Then, modify the constructor for shorthand objects so that you can pass in the accumulated dictionary as `**kwargs`.

Ideally, the initial values would also be determined once at startup, rather than dynamically. They're not going to change, so there's no reason to make lookup dynamic.

That said: is there any reason to go for the initial values at all? If you haven't overwritten any font properties, isn't the shortand font declaration "don't declare anything" - i.e. an empty declaration? I think it would be surprising for 
a declaration to go from "normal normal normal 12pt/system..." to "italic" because you set a single property.

Reply:
>My suggestion: use a dictionary, rather than the shorthand object as an accumulator. This means the "was anything manually set" check is an "if {}" check. Then, modify the constructor for shorthand objects so that you can pass in the accumulated dictionary as **kwargs.

Yes this makes sense, thanks

>Ideally, the initial values would also be determined once at startup, rather than dynamically. They're not going to change, so there's no reason to make lookup dynamic.
>That said: is there any reason to go for the initial values at all?

If when asking for the shorthand property returning an empty string is fine with you, then no it is not needed.

Reply:
Empty string or None would both be fine in my book. My reasoning is that the "reversible" state of getting a shortcut description isn't "explicitly set all the component attributes to their defaults", it's "revert all the component values to their defaults", so any logic that makes it easy to identify that case is fine.

Reply:
Fixed